ement-room.el

```diff
@@ -2817,151 +2809,113 @@ MSG is the message STRING was contained in."
-                              ((string-match-p "book.*" type) "Book")
-                              ((string-match-p "profile.*" type) "Profile"))))
-      (setf string (concat string
-                           (propertize typestr 'face 'ement-room-preview)
-                           "\n")))
-    (when-let ((desc (alist-get 'og:description data)))
-      (let* ((current-left-margin (or (car (window-margins)) 0))
-             (current-right-margin (or (cdr (window-margins)) 0))
-             (available-width (abs (- (- (window-width)
-                                         (+ current-left-margin current-right-margin))
-                                      (length ement-room-url-preview-prefix)))))
-        (setf string (concat string
-                             (propertize (truncate-string-to-width (replace-regexp-in-string "\n"
-                                                                                             " "
-                                                                                             desc)
-                                                                   available-width)
-                                         'face 'ement-room-preview)
-                             "\n"))))
-    (propertize string 'line-prefix ement-room-url-preview-prefix)))
+  (with-temp-buffer
+    (when-let ((title (alist-get 'og:title data)))
+      (insert (propertize title
+                          'mouse-face 'highlight
+                          'face '(link (:inherit ement-room-preview))
+                          'help-echo (alist-get 'og:url data)
+                          'action (lambda (d)
+                                    (when (stringp d)
+                                      (browse-url d)))
+                          'follow-link (alist-get 'og:url data)
+                          'button t
+                          'button-data (alist-get 'og:url data)
+                          'keymap button-map
+                          'category t)
+              "\n"))
+    (let* ((type? (alist-get 'og:type data))
+           (special? (alist-get 'og:site_name data))
+           (typestr nil))
+      (if special?
+          (setf typestr special?)
+        (setf typestr (cond ((not type?) nil)
+                            ((string-match-p "music.*" type?) "Music")
+                            ((string-match-p "video.*" type?) "Video")
+                            ((string-match-p "article.*" type?) "Article")
+                            ((string-match-p "book.*" type?) "Book")
+                            ((string-match-p "profile.*" type?) "Profile"))))
+      (when typestr
+        (insert (propertize typestr 'face 'ement-room-preview) "\n")))
+    (when-let* ((desc (alist-get 'og:description data))
+                (current-left-margin (or (car (window-margins)) 0))
+                (current-right-margin (or (cdr (window-margins)) 0))
+                (available-width (abs (- (- (window-width)
```

Comment:
I don't think this code regarding window width and margins should be necessary.  Emacs should handle that kind of thing automatically.  As well, margins may not be used, depending on the user's configuration.  So let's keep this code as simple as possible.

Reply:
It's mostly just to keep the description a reasonable length. Some sites with poorly implemented previews have paragraphs of text as their description, which would all get inserted unless it's cropped :)

It currently already handles the case where margins are not used (it just uses left and right as 0).

A probably simpler and maybe better way though is to define a customization option for max length of descriptions and just truncate all descriptions to that length, but the window-width way is nice in that it is adaptive to the current size of the ement buffer, so for example if you have the ement room fullscreen as the only window, it will show enough preview of the description to fill that space, and it will get smaller as you decrease the size of the window.

I totally understand if you don't want to complicate things though, so I'm fine with whatever one you prefer.